fsspec/spec.py

```diff
@@ -428,11 +428,11 @@ def walk(self, path, maxdepth=None, topdown=True, on_error="omit", **kwargs):
         except (FileNotFoundError, OSError) as e:
             if on_error == "raise":
                 raise
-            elif callable(on_error):
+            if callable(on_error):
                 on_error(e)
-            if detail:
-                return path, {}, {}
-            return path, [], []
+            result = (path, {}, {}) if detail else (path, [], [])
+            yield result
```

Comment:
Am I correct that we want both?
1. Yield an empty result (`yield`).
2. Raise `StopIteration` (`return`).

Reply:
Correct, the generator should stop after yielding the single tuple. I don't know why `return` should not be the simple way to achieve that...

Reply:
Testing this, it seems you are right, and the original code is equivalent to `raise StopIteration` by itself with no value at all. 

```python
In [4]: def genme():
   ...:     if True:
   ...:         return 1
   ...:     yield 2
   ...:

In [5]: list(genme())
Out[5]: []
```

Reply:
Here's the error:
```
  Full diff:
  - []
  + [
  +     (
  +         '/do_not_exist',
  +         [],
  +         [],
  +     ),
  + ]
```
Actually, I think we do not want to yield anything. So `return` was correct. However, I don't understand why `return path, [], []` instead of a mere `return`.

Reply:
Before we were returning nothing (even though the return statement had a value), now we yield the tuple first.